### ISSUE TITLE
🏗️ Architectural Refactor: Implement Secure and Dynamic Position Sizing

### ISSUE THREAD

**Issue by SefaGH** [SefaGH/bearish-alpha-bot#290]
about: "Critical architectural refactor to fix 95% signal rejection rate"
labels: enhancement, critical, architecture, bug
---

**Reported by:** SefaGH  
**Date:** 2025-11-05 10:10:22 UTC  
**Priority:** 🔴 CRITICAL  
**Target Completion:** 2025-11-19  

### **Executive Summary**
Bot generates valid trading signals but **95% are rejected** by RiskManager due to position size exceeding capital limits ($133 position on $100 capital).

### **Root Cause**
- RiskManager calculates AND validates (dual responsibility violation)
- Position sizing formula lacks capital protection caps
- Configuration uses hardcoded values instead of GitHub Variables

### **New Architecture Flow**
```
Signal → Coordinator → PositionSizing(calculate) → RiskManager(validate) → Execute
```

### **Instructions for Copilot**
Please implement the following changes in order. Use the existing configuration management system with GitHub Variables priority.

---

### **✅ Task 1: Update Configuration Management**

**File:** `config/config.example.yaml`

Update the risk management section to support new dynamic sizing parameters. These will be overrideable by GitHub Variables:

```yaml
# ==============================================================================
# RISK MANAGEMENT - ENHANCED POSITION SIZING
# ==============================================================================
risk:
  # Existing parameters (keep these)
  equity_usd: 100                     # Override with: CAPITAL_USDT
  per_trade_risk_pct: 0.01            # Override with: PER_TRADE_RISK_PCT
  daily_loss_limit_pct: 0.02          # Override with: DAILY_LOSS_LIMIT_PCT
  risk_usd_cap: 5                     # Override with: RISK_USD_CAP
  
  # NEW: Dynamic position sizing parameters
  max_position_size_pct: 0.20         # Override with: MAX_POSITION_SIZE_PCT
  max_notional_pct_per_trade: 0.20    # Override with: MAX_NOTIONAL_PCT_PER_TRADE
  max_margin_pct_per_trade: 0.20      # Override with: MAX_MARGIN_PCT_PER_TRADE
  
  # Safety limits
  min_position_size_usd: 10           # Override with: MIN_POSITION_SIZE_USD
  min_stop_pct: 0.003                 # Override with: MIN_STOP_PCT
  
  # Remove deprecated (if exists)
  # max_notional_per_trade: 20        # REMOVE THIS - causes the bug
```

---

### **✅ Task 2: Fix live_trading_launcher.py Configuration Mapping**

**File:** `scripts/live_trading_launcher.py`

Locate the `_normalize_risk_params()` method and fix the incorrect mapping that causes the bug:

```python
def _normalize_risk_params(self):
    """
    CRITICAL FIX: Remove max_notional_per_trade from max_position_size mapping
    """
    if not hasattr(self, 'RISK_PARAMS'):
        self.RISK_PARAMS = {}
    
    # FIXED mapping - DO NOT map max_notional_per_trade to max_position_size!
    key_mappings = {
        'max_position_size': ['max_position_size_pct'],  # REMOVED 'max_notional_per_trade'
        'stop_loss_pct': ['stop_loss_pct'],              # REMOVED 'min_stop_pct'
        'take_profit_pct': ['take_profit_pct'],
        'min_stop_pct': ['min_stop_pct'],                # Separate parameter
        'risk_per_trade': ['per_trade_risk_pct'],
        'max_drawdown': ['daily_loss_limit_pct'],
        # NEW mappings for dynamic sizing
        'max_notional_pct': ['max_notional_pct_per_trade'],
        'max_margin_pct': ['max_margin_pct_per_trade'],
    }
    
    # ... existing mapping logic ...
    
    # CRITICAL: Add safety check for max_position_size
    if 'max_position_size' in self.RISK_PARAMS:
        value = self.RISK_PARAMS['max_position_size']
        if value > 1.0:
            logger.warning(f"⚠️ max_position_size={value} > 1.0, converting to ratio")
            self.RISK_PARAMS['max_position_size'] = min(value / 100.0, 0.20)  # Cap at 20%
            logger.info(f"✅ Corrected max_position_size to {self.RISK_PARAMS['max_position_size']:.2%}")
```

---

### **✅ Task 3: Enhance AdvancedPositionSizing Module**

**File:** `src/core/position_sizing.py`

Add the new `calculate_optimal_size` method with two-stage safety calculation:

```python
class AdvancedPositionSizing:
    """
    Advanced position sizing with dynamic capital-based caps.
    """
    
    async def calculate_optimal_size(self, signal: Dict, method: str = 'fixed_risk_capped') -> Dict:
        """
        Calculate position size with two-stage safety:
        1. Risk-based calculation
        2. Dynamic capital percentage caps
        
        This method enriches the signal with:
        - amount: Position size in base currency
        - notional: Position value in USDT
        - sizing_meta: Detailed calculation metadata
        """
        try:
            # Get required parameters
            symbol = signal.get('symbol', 'UNKNOWN')
            entry_price = signal.get('entry', 0) or signal.get('entry_price', 0)
            stop_loss = signal.get('stop', 0) or signal.get('stop_loss', 0)
            
            # Calculate stop loss percentage
            if entry_price > 0 and stop_loss > 0:
                stop_pct = abs(entry_price - stop_loss) / entry_price
            else:
                logger.warning(f"[SIZING] Missing price data for {symbol}")
                return signal
            
            # Get portfolio state from RiskManager
            portfolio = self.risk_manager.get_portfolio_summary()
            capital = float(portfolio.get('portfolio_value', 0))
            
            if capital <= 0:
                logger.error(f"[SIZING] Invalid capital: {capital}")
                return signal
            
            # Get configuration (use config from risk_manager)
            config = getattr(self.risk_manager, 'config', {})
            
            # Use configuration with GitHub Variables priority
            risk_pct = float(config.get('per_trade_risk_pct', 0.01))
            risk_cap = config.get('risk_usd_cap')
            max_notional_pct = float(config.get('max_notional_pct_per_trade', 0.20))
            max_margin_pct = float(config.get('max_margin_pct_per_trade', 0.20))
            leverage = float(signal.get('leverage', config.get('leverage_default', 5)))
            
            # STAGE 1: Risk-based calculation
            base_risk_usd = capital * risk_pct
            if risk_cap:
                base_risk_usd = min(base_risk_usd, float(risk_cap))
            
            risk_based_notional = base_risk_usd / stop_pct if stop_pct > 0 else 0
            
            # STAGE 2: Apply dynamic caps (percentage of capital)
            exposure_cap = capital * max_notional_pct
            margin_cap = capital * max_margin_pct * leverage  # For futures
            
            # Determine final cap
            caps_to_apply = [exposure_cap, margin_cap]
            final_cap = min(caps_to_apply)
            
            # Apply the cap
            final_notional = min(risk_based_notional, final_cap)
            
            # Calculate position amount
            amount = final_notional / entry_price if entry_price > 0 else 0
            
            # Create detailed metadata
            sizing_meta = {
                'method': method,
                'capital': capital,
                'risk_pct': risk_pct,
                'stop_pct': stop_pct,
                'calculations': {
                    'base_risk_usd': base_risk_usd,
                    'risk_based_notional': risk_based_notional,
                    'exposure_cap': exposure_cap,
                    'margin_cap': margin_cap,
                    'final_notional': final_notional
                },
                'position_pct': (final_notional/capital)*100 if capital > 0 else 0,
                'capped': risk_based_notional > final_cap
            }
            
            # Log the calculation
            logger.info(f"📊 [SIZING] {symbol}")
            logger.info(f"   Capital: ${capital:.2f}")
            logger.info(f"   Risk-based: ${risk_based_notional:.2f}")
            logger.info(f"   Cap applied: ${final_cap:.2f}")
            logger.info(f"   Final notional: ${final_notional:.2f} ({sizing_meta['position_pct']:.1f}% of capital)")
            
            # Enrich signal with calculated values
            signal['amount'] = amount
            signal['notional'] = final_notional
            signal['position_size'] = amount  # Backward compatibility
            signal['leverage'] = leverage
            signal['sizing_meta'] = sizing_meta
            
            return signal
            
        except Exception as e:
            logger.error(f"[SIZING] Error calculating size for {signal.get('symbol')}: {e}", exc_info=True)
            return signal
```

---

### **✅ Task 4: Update StrategyCoordinator Integration**

**File:** `src/core/strategy_coordinator.py`

Update the `_assess_signal_risk` method to use AdvancedPositionSizing:

```python
async def _assess_signal_risk(self, signal: Dict) -> Dict[str, Any]:
    """
    UPDATED: Use AdvancedPositionSizing instead of risk_manager for sizing.
    """
    try:
        # Initialize position sizing if not already done
        if not hasattr(self, 'position_sizing'):
            from core.position_sizing import AdvancedPositionSizing
            self.position_sizing = AdvancedPositionSizing(self.risk_manager)
            logger.info("✅ AdvancedPositionSizing initialized in StrategyCoordinator")
        
        # Calculate position size using the new module
        sized_signal = await self.position_sizing.calculate_optimal_size(signal)
        
        # Check if sizing was successful
        if sized_signal.get('amount', 0) <= 0:
            return {
                'acceptable': False,
                'reason': 'Unable to calculate valid position size',
                'metrics': sized_signal.get('sizing_meta', {})
            }
        
        # NOW validate the sized position with risk rules
        is_valid, reason, risk_metrics = await self.risk_manager.validate_new_position(
            sized_signal,
            portfolio_manager=self.portfolio_manager
        )
        
        if not is_valid:
            return {
                'acceptable': False,
                'reason': reason,
                'metrics': risk_metrics
            }
        
        # Add sizing metadata to risk metrics
        risk_metrics['sizing_meta'] = sized_signal.get('sizing_meta', {})
        
        return {
            'acceptable': True,
            'position_size': sized_signal['amount'],
            'notional': sized_signal['notional'],
            'metrics': risk_metrics
        }
        
    except Exception as e:
        logger.error(f"Error in risk assessment: {e}", exc_info=True)
        return {
            'acceptable': False,
            'reason': f"Risk assessment error: {str(e)}",
            'metrics': {}
        }
```

---

### **✅ Task 5: Update Risk Rules for Margin Support**

**File:** `src/core/risk_rules.py`

Update the `CapitalLimitRule` class to support margin-based validation:

```python
class CapitalLimitRule(BaseRiskRule):
    """
    UPDATED: Now supports margin-based validation for futures.
    - Spot (leverage=1): Compare notional to available capital
    - Futures (leverage>1): Compare required margin to available capital
    """
    
    def validate(self, signal: Dict, portfolio_manager) -> Tuple[bool, str]:
        """
        Validate capital requirements based on leverage.
        """
        try:
            symbol = signal.get('symbol', 'UNKNOWN')
            notional = signal.get('notional', 0)
            leverage = signal.get('leverage', 1)
            
            # Get portfolio state
            portfolio = portfolio_manager.get_portfolio_summary()
            available = portfolio.get('available_capital', portfolio.get('portfolio_value', 0))
            
            if notional <= 0:
                return (False, f"Invalid notional value: {notional}")
            
            # Spot trading (leverage = 1)
            if leverage <= 1:
                if notional > available:
                    logger.warning(f"🚫 [CapitalLimitRule] REJECTED {symbol} (spot): ${notional:.2f} > ${available:.2f}")
                    return (False, f"Spot position ${notional:.2f} exceeds available ${available:.2f}")
                else:
                    logger.info(f"✅ [CapitalLimitRule] PASSED {symbol} (spot): ${notional:.2f} <= ${available:.2f}")
                    return (True, "Capital check passed (spot)")
            
            # Futures trading (leverage > 1)
            required_margin = notional / leverage
            
            if required_margin > available:
                logger.warning(f"🚫 [CapitalLimitRule] REJECTED {symbol} (futures): Margin ${required_margin:.2f} > ${available:.2f}")
                return (False, f"Margin ${required_margin:.2f} exceeds available ${available:.2f}")
            else:
                logger.info(f"✅ [CapitalLimitRule] PASSED {symbol} (futures): Margin ${required_margin:.2f} <= ${available:.2f}")
                return (True, f"Margin check passed (leverage {leverage}x)")
                
        except Exception as e:
            logger.error(f"[CapitalLimitRule] Error: {e}", exc_info=True)
            return (False, f"Validation error: {str(e)}")
```

---

## 4. 📊 GitHub Variables to Create

**@SefaGH - Please create these Repository Variables in GitHub:**

Navigate to: `Settings → Secrets and variables → Actions → Variables tab`

### **Required New Variables**

| Variable Name | Default Value | Description | Priority |
|--------------|---------------|-------------|----------|
| `MAX_POSITION_SIZE_PCT` | `0.20` | Maximum position size as % of capital (0.20 = 20%) | HIGH |
| `MAX_NOTIONAL_PCT_PER_TRADE` | `0.20` | Maximum notional per trade as % of capital | HIGH |
| `MAX_MARGIN_PCT_PER_TRADE` | `0.20` | Maximum margin usage per trade as % of capital | HIGH |
| `MIN_POSITION_SIZE_USD` | `10` | Minimum position size in USD (exchange limit) | MEDIUM |
| `LEVERAGE_DEFAULT` | `5` | Default leverage for futures trading | MEDIUM |
| `LEVERAGE_MAX` | `20` | Maximum allowed leverage | LOW |

### **Variables to REMOVE or Update**

| Variable Name | Action | Reason |
|--------------|--------|---------|
| `MAX_NOTIONAL_PER_TRADE` | **REMOVE** | Causes the bug - being mapped incorrectly |
| `MIN_STOP_PCT` | Keep but ensure not mapped to `stop_loss_pct` | Different purpose |

### **Configuration Priority Reminder**
```
1. GitHub Variables (HIGHEST)
2. Environment Variables (.env)
3. config.yaml
4. Hardcoded defaults (LOWEST)
```

---

## 5. 🧪 Validation Tests

After implementation, run these tests to verify the fix:

```python
# Test 1: Normal position sizing
capital = 100
stop_pct = 0.01  # 1%
expected_notional = 20  # Capped at 20% of capital
assert calculated_notional == 20

# Test 2: Tight stop loss scenario (the bug case)
capital = 100
stop_pct = 0.001  # 0.1% - would give $1000 position
expected_notional = 20  # Should be capped at 20%
assert calculated_notional == 20

# Test 3: Futures with leverage
capital = 100
leverage = 10
margin_cap = 20  # 20% of capital for margin
expected_margin = 2  # $20 notional / 10x leverage
assert required_margin <= margin_cap
```

---

## 6. 📋 Copilot Action Checklist

**Please complete these tasks in order:**

- [ ] Update `config/config.example.yaml` with new risk parameters
- [ ] Fix `scripts/live_trading_launcher.py` - `_normalize_risk_params()` method
- [ ] Update `src/core/position_sizing.py` - Add `calculate_optimal_size()` method
- [ ] Update `src/core/strategy_coordinator.py` - Modify `_assess_signal_risk()` method
- [ ] Update `src/core/risk_rules.py` - Enhance `CapitalLimitRule` class
- [ ] Run existing tests to ensure no regression
- [ ] Add logging statements as shown in the code examples

**Success Criteria:**
- Position sizes never exceed 20% of capital
- Margin-based validation works for futures
- All existing tests still pass
- Detailed sizing logs are generated

---

## 7. 🚀 Expected Outcome

After these changes:
- **Rejection rate:** 95% → <15%
- **Position sizing:** $133 on $100 capital → $20 on $100 capital (max 20%)
- **Daily trades:** 0-1 → 5-10
- **Architecture:** Clean separation of sizing vs validation

---

**Please proceed with the implementation following the centralized configuration management pattern. Use GitHub Variables as the primary configuration source.**
